backend/src/main/java/com/example/backend/Specifications/FlightSpecifications.java

```diff
@@ -72,4 +75,97 @@ public static Specification<Flight> hasDepartureDate(LocalDate departureDate) {
     public static Specification<Flight> hasArrivalDateLessThan(LocalDate arrivalDate) {
         return (root, query, criteriaBuilder) -> criteriaBuilder.lessThan(root.get("arrivalDate"), arrivalDate);
     }
+
+    public static Specification<Flight> hasDepartureAirport(Integer id) {
+        return (root, query, criteriaBuilder) -> {
+            Join<Flight, FlightLeg> flightAndFlightLeg = root.join("flightLegs");
+            Join<FlightLeg, Airport> flightLegAndAirport = flightAndFlightLeg.join("departureAirport");
+            return criteriaBuilder.and(
+                    criteriaBuilder.equal(flightLegAndAirport.get("id"), id),
+                    criteriaBuilder.equal(flightAndFlightLeg.get("flightLegId"), 1));
+        };
+    }
+
+    public static Specification<Flight> hasArrivalAirport(Integer id) {
+        return (root, query, criteriaBuilder) -> {
+
+            if (query == null) {
+                throw new IllegalArgumentException("Query parameter cannot be null");
+            }
+
+            Subquery<Integer> subquery = createNumberOfLastFlightLegSubquery(root, query, criteriaBuilder);
+
+            Join<Flight, FlightLeg> flightAndFlightLeg = root.join("flightLegs");
+            Join<FlightLeg, Airport> flightLegAndAirport = flightAndFlightLeg.join("arrivalAirport");
+
+            return criteriaBuilder.and(
+                    criteriaBuilder.equal(flightAndFlightLeg.get("flightLegId"), subquery),
+                    criteriaBuilder.equal(flightLegAndAirport.get("id"), id));
+        };
+    }
+
```

Comment:
For the hasDepartureAirport and hasArrivalAirport functions, they could be just one function with extra parameter to prevent duplicate code.